### PR TITLE
Special processing for dotfiles

### PR DESCRIPTION
**Pull request by jvkersch**
This is a PR out of the patch that I [submitted to stow-devel](http://lists.gnu.org/archive/html/stow-devel/2016-05/msg00000.html) to allow for special processing for dotfiles. I use GNU stow to manage my dotfiles, but found it a bit hard to maintain an apparently empty Stow folder, which is full of hidden files.

To address this, the idea is to have a command-line option `--dotfiles` that would have Stow check if the files/folders to stow start with `dot-` (though I should make the prefix configurable...). In this case, Stow replaces the `dot-` prefix by an actual dot (`.`), turning the link target into a proper hidden file.

I am not a skilled Perl developer, so the code is probably rough around the edges. However, if this is a viable approach then I'm definitely up for polishing it and bringing it up to standards.

One corner case that I haven't explored yet is the interplay between this kind of processing and folding. I imagine that the dotfiles option would disable folding, but perhaps there is a way to have both.

### CODE REVIEW

File: lib/Stow/Util.pm.in [hunk 1]
```diff
@@ -193,6 +193,20 @@ sub restore_cwd {
     chdir($prev) or error("Your current directory $prev seems to have vanished");
 }
 
+sub adjust_dotfile {
+    my $target = $_[0];
+
+    my @result = ();
+    for my $part (split m{/+}, $target) {
+        if ($part ne "dot-") {
```
Comment (aspiers) [inline lib/Stow/Util.pm.in]: I like the attention to detail here :-)  I think you need to also defend against `dot-.` though, as that would currently be turned into `..`.

File: lib/Stow/Util.pm.in [hunk 2]
```diff
@@ -193,6 +193,20 @@ sub restore_cwd {
     chdir($prev) or error("Your current directory $prev seems to have vanished");
 }
 
+sub adjust_dotfile {
+    my $target = $_[0];
```
Comment (aspiers) [inline lib/Stow/Util.pm.in]: my ($target) = @_;

would be more consistent with the result of Stow's code style.

File: lib/Stow.pm.in [hunk 3]
```diff
@@ -377,6 +378,13 @@ sub stow_contents {
         next NODE if $node eq '..';
         my $node_target = join_paths($target, $node);
         next NODE if $self->ignore($stow_path, $package, $node_target);
+
+	if ($self->{dotfiles}) {
```
Comment (aspiers) [inline lib/Stow.pm.in]: Please use spaces to indent, not tabs, to stay consistent with the rest of the code base.

File: lib/Stow.pm.in [hunk 4]
```diff
@@ -744,6 +753,13 @@ sub unstow_contents {
...
```
Comment (aspiers) [inline lib/Stow.pm.in]: Use spaces not tabs here too.

File: lib/Stow.pm.in [hunk 5]
```diff
@@ -801,6 +817,12 @@ sub unstow_node {
         # Does the existing $target actually point to anything?
         if (-e $existing_path) {
             # Does link points to the right place?
+
+	    # Adjust for dotfile if necessary.
```
Comment (aspiers) [inline lib/Stow.pm.in]: Use spaces not tabs here too.

File: lib/Stow.pm.in [hunk 6]
```diff
@@ -617,6 +625,7 @@ sub unstow_contents_orig {
         next NODE if $node eq '..';
         my $node_target = join_paths($target, $node);
         next NODE if $self->ignore($stow_path, $package, $node_target);
+        # $node_target = adjust_dotfile($node_target);
```
Comment (aspiers) [inline lib/Stow.pm.in]: We don't want this line, but it will be cancelled out by the next commit when you squash the two commits into one anyway, so that's OK.

File: t/dotfiles.t [hunk 7]
```diff
@@ -0,0 +1,72 @@
+#!/usr/local/bin/perl
```
Comment (aspiers) [inline t/dotfiles.t]: When adding a new file, you also need to update `MANIFEST`. That's the reason that the Travis CI builds are currently failing for this pull request.

**Comment by jvkersch** [comment]
@aspiers Thanks for the insightful comments! I think I've addressed them all, but there might be other things still missing (for instance, I should write some docs, and do something about the decrease in test coverage).

File: lib/Stow.pm.in [hunk 8]
```diff
@@ -744,6 +752,13 @@ sub unstow_contents {
         next NODE if $node eq '..';
         my $node_target = join_paths($target, $node);
         next NODE if $self->ignore($stow_path, $package, $node_target);
+
+    if ($self->{dotfiles}) {
```
Comment (aspiers) [inline lib/Stow.pm.in]: Wrong indentation here.  I'll see if GitHub lets me correct it myself though.

**Comment by aspiers** [comment]
Hmm no, it seems that you have to enable the [`Allow edits from maintainers` option](https://github.com/blog/2247-improving-collaboration-with-forks) before GitHub will let me push to your fork.  So please can you either fix that indentation issue, or enable that option?  Thanks!

**Comment by jvkersch** [comment]
@aspiers Thanks for the swift feedback and sorry for the oversight. I've fixed it (and enabled the option for future use, if necessary). I also added some more tests for some corner cases in the dot expansion, and I wrote a small summary of what the option does.

One remaining sore spot is the decrease in coverage. I need to figure out what lines aren't being hit, and then write a separate test for them. I'll do so as soon as possible.

**Comment by jvkersch** [comment]
@aspiers I added a simple unstow test. Please let me know if you want to see any further changes, and thanks again for all the reviewing.

**Comment by lassik** [comment]
This feature uses `dot-` as a prefix for dotfile directories. At least in BSD-land the usual prefix is `dot.` Examples:

* https://github.com/openbsd/src/tree/master/etc/root
* https://github.com/freebsd/freebsd/tree/master/share/skel

I also found these files on my Mac: `dot.mg` and `dot.gdbinit`.

On the other hand, in one Emacs package I found `dot-emacs.el`.

On Alpine Linux or Ubuntu I didn't find any files with either prefix.

**Comment by paride** [comment]
FWIW, I would favor the `dot.` prefix.

**Comment by shajra** [comment]
It seems to be that there could be too-open debate on what the right prefix should be.  Would it be better coming in as a configuration parameter optionally, with a reasonable default?
